### PR TITLE
KT-45777: Allow 2 levels of granularity when tracking changes

### PR DESCRIPTION
**Pull request by hungvietnguyen**

### CODE REVIEW

File: compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/ClasspathSnapshotShrinker.kt [hunk 1]
```diff
@@ -105,15 +104,26 @@ object ClasspathSnapshotShrinker {
...
-        val referencedClassIds = referencedClasses.map { it.classSnapshot.getClassId() }.toSet()
+        val referencedClassIds = referencedClasses.mapTo(mutableSetOf()) { it.classId }
         val transitivelyReferencedClassIds: Set<ClassId> =
             ImpactAnalysis.findImpactedClassesInclusive(referencedClassIds, supertypesResolver) // Use Set for presence check
 
-        return allClasses.filter { it.classSnapshot.getClassId() in transitivelyReferencedClassIds }
+        return allClasses.filter { it.classId in transitivelyReferencedClassIds }
+    }
+
+    class MetricsReporter(
```
Comment (hungvietnguyen) [inline compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/ClasspathSnapshotShrinker.kt:118]: Ivan's comment from https://github.com/google/kotlin/commit/38227b5843d51941f80b8b96585ef74fd363e860#r64567154

> "IMHO using the same pattern everywhere (BuildMetricsReporter directly) is better, for consistency."

Right, this pattern is not great. But I created this because this `ClasspathSnapshotShrinker` is called in 2 places (before and after the compilation), so the `BuildTime` enums can't be shared. So I need a way to force the caller to provide a list of `BuildTime`s based on the caller's context and pass it to the shrinker.

File: build-common/src/org/jetbrains/kotlin/incremental/storage/externalizers.kt [hunk 2]
```diff
@@ -375,27 +375,34 @@ class ListExternalizer<T>(elementExternalizer: DataExternalizer<T>) :
...
         val size = input.readInt()
-        val map = LinkedHashMap<K, V>(size)
+        val map = newMap(size)
         repeat(size) {
             val key = keyExternalizer.read(input)
             val value = valueExternalizer.read(input)
             map[key] = value
         }
-        return map
+        @Suppress("UNCHECKED_CAST")
```
Comment (hungvietnguyen) [inline build-common/src/org/jetbrains/kotlin/incremental/storage/externalizers.kt:400]: Ivan's comment from https://github.com/google/kotlin/commit/38227b5843d51941f80b8b96585ef74fd363e860#r64563620

> To make it a bit more type-safe, maybe about having default value for `newMap` and change it to `(Int) -> M`?

I just tried it and `map[key] = value` on line 398 above will not compile because `M` is an immutable map.

Comment (gavra0) [reply build-common/src/org/jetbrains/kotlin/incremental/storage/externalizers.kt:400]: Does changing to `<M: MutableMap<K, V>>` work?

File: compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/ClasspathSnapshot.kt [hunk 3]
```diff
@@ -26,7 +26,7 @@ class ClasspathEntrySnapshot(
      * Maps (Unix-style) relative paths of classes to their snapshots. The paths are relative to the containing classpath entry (directory
      * or jar).
      */
-    val classSnapshots: LinkedHashMap<String, ClassSnapshotWithHash>
+    val classSnapshots: LinkedHashMap<String, ClassSnapshot>
```
Comment (gavra0) [inline compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/ClasspathSnapshot.kt:29]: Is this change intentional? This makes it possible to have `InaccessibleClassSnapshot` here.

**Review by gavra0 — COMMENTED** [review]
Can you please add unit tests for the new snapshotting and comparing code?

File: compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/ClasspathSnapshot.kt (hunk 3 above)
Comment (hungvietnguyen) [reply compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/ClasspathSnapshot.kt:29]: Yes it is intended. 

We need to remove duplicates first before we can remove `InaccessibleClassSnapshot` (there will be a bug otherwise--as detailed below). And removing duplicates requires looking at the entire original classpath (where nothing is removed).

I wrote some kdoc in the method below, quoting it here:
```
/**
 * Removes duplicate classes and [InaccessibleClassSnapshot]s from the given [ClasspathSnapshot].
 *
 * To see why removing duplicate classes is important, consider this example:
 *   - Current classpath: (Unchanged) jar2!/com/example/A.class containing A.foo, (Added) jar3!/com/example/A.class containing A.bar
 *   - Previous classpath: (Removed) jar1!/com/example/A.class containing A.bar, (Unchanged) jar2!/com/example/A.class containing A.foo
 * Without removing duplicates, we might report that there are no changes (both the current classpath and previous classpath have A.foo and
 * A.bar). However, the correct report should be that A.bar is removed and A.foo is added because the second A class on each classpath does
 * not have any effect.
 *
 * It's also important to remove duplicate classes first before removing [InaccessibleClassSnapshot]s. For example, if
 * jar1!/com/example/A.class is inaccessible and jar2!/com/example/A.class is accessible, removing inaccessible classes first would mean
 * that jar2!/com/example/A.class would be kept whereas it shouldn't be since it is a duplicate class (keeping a duplicate class can
 * lead to incorrect change reports as shown in the previous example).
 *
 * That is also why we cannot remove inaccessible classes from each classpath entry in isolation (i.e., during classpath entry
 * snapshotting), even though it seems more efficient to do so. For correctness, we need to look at the entire classpath first, remove
 * duplicate classes, and then remove inaccessible classes.
 */
internal fun ClasspathSnapshot.removeDuplicateAndInaccessibleClasses(): List<AccessibleClassSnapshot> {
    return getNonDuplicateClassSnapshots().filterIsInstance<AccessibleClassSnapshot>()
}
```

**Comment by hungvietnguyen** [comment]
> Can you please add unit tests for the new snapshotting and comparing code?

Sure thing, I had a note for myself on this (the PR was a bit large so I didn't add it too early, in case we change directions).

File: build-common/src/org/jetbrains/kotlin/incremental/storage/externalizers.kt (hunk 2 above)
Comment (gavra0) [reply build-common/src/org/jetbrains/kotlin/incremental/storage/externalizers.kt:400]: I think having type-safety is not enforcing mutability/immutability of the structure you are serializing/de-serializing. If `M: MutableMap<K, V>>` this means that `M` is a subtype of `Map<K, V>` so you could maybe use that if you want.

File: compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/ClasspathSnapshot.kt (hunk 3 above)
Comment (gavra0) [reply compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/ClasspathSnapshot.kt:29]: Ah right, for some reason I thought that `InaccessibleClassSnapshot` cannot be converted to `ClassSnapshotWithHash`, but that is incorrect.

File: build-common/src/org/jetbrains/kotlin/incremental/storage/externalizers.kt (hunk 2 above)
Comment (hungvietnguyen) [reply build-common/src/org/jetbrains/kotlin/incremental/storage/externalizers.kt:400]: Okay so let me clarify my intentions first.

Suppose we need an externalizer for `val data: Map<String, String>`. My goal is for the usage to look like this:

```
val x: Map<String, String> = MapExternalizer(StringExternalizer, StringExternalizer).read(input) // compiles okay
val y: MutableMap<String, String> = MapExternalizer(StringExternalizer, StringExternalizer).read(input) // fails to compile
```

That is, the `read` method should return an immutable map.

The current implementation of `MapExternalizer` achieves that goal (even though I agree that the unsafe casting is not great).

I tried your suggestions, but it required removing the default value of `newMap`:
```
open class MapExternalizer<K, V, M : MutableMap<K, V>>(
    ...
    private val newMap: (size: Int) -> M // Removed: MutableMap<K, V> = { size -> LinkedHashMap(size) }
}    
```    

So the usage will look like this:
```
val x: Map<String, String> = MapExternalizer(StringExternalizer, StringExternalizer, {LinkedHashMap()}).read(input) // compiles okay
val y: MutableMap<String, String> = MapExternalizer(StringExternalizer, StringExternalizer, {LinkedHashMap()}).read(input) // compiles okay
```

There are 2 effects at the usage:
  - I have to provide the extra `{LinkedHashMap()}` argument, which should be an internal detail of `MapExternalizer`.
  - The `read` method returns a mutable map now.
  
If there is an alternative implementation that achieves my previous goal, that would be great.

Comment (gavra0) [reply build-common/src/org/jetbrains/kotlin/incremental/storage/externalizers.kt:400]: Is `LinkedHashMapExternalizer` the only implementation? If so there is no need for this base class.

Also, please note that you can solve this issue by creating parametrized subclasses of `MapExternalizer` with concrete `newMap` params, as you've done in `LinkedHashMapExternalizer`.

Comment (hungvietnguyen) [reply build-common/src/org/jetbrains/kotlin/incremental/storage/externalizers.kt:400]: At one point during implementation, I had this two classes:
```
class ClasspathEntrySnapshot(
    val classSnapshots: **LinkedHashMap**<String, ClassSnapshot>
)

class PackageFacadeKotlinClassSnapshot(
    val packageMembers: **Map**<FqName, String>
) : KotlinClassSnapshot()
```

So I created `MapExternalizer` and let `LinkedHashMapExternalizer` extend it.

But later on, I changed `PackageFacadeKotlinClassSnapshot` to:
```
class PackageFacadeKotlinClassSnapshot(
    val packageMemberNames: **Set**<String>
) : KotlinClassSnapshot()
```

So `MapExternalizer` is unused now (except by `LinkedHashMapExternalizer`). I decided to keep it because:
  - The use of `Map` is common, and we'll have `MapExternalizer` already available when we need it.
  - `ListExternalizer` and `SetExternalizer` extend `GenericCollectionExternalizer` (which btw has the same unsafe typecast issue), so if we have `LinkedHashMapExternalizer` extend `MapExternalizer`, it will be the same pattern.
  
That said, I do agree `MapExternalizer` is currently unused so I can also remove it. (The only issue is that if I have `val data: Map<X, Y>` in the future, I will need to introduce `MapExternalizer` back again -- which is fine.)

WDYT?

Comment (gavra0) [reply build-common/src/org/jetbrains/kotlin/incremental/storage/externalizers.kt:400]: Removing it, and having just a single implementation, sounds good to me.

Comment (hungvietnguyen) [reply build-common/src/org/jetbrains/kotlin/incremental/storage/externalizers.kt:400]: Sure thing, I'll fix that.

**Comment by hungvietnguyen** [comment]
> > Can you please add unit tests for the new snapshotting and comparing code?
> 
> Sure thing, I had a note for myself on this (the PR was a bit large so I didn't add it too early, in case we change directions).

I've added the tests, PTAL.

File: compiler/incremental-compilation-impl/test/org/jetbrains/kotlin/incremental/classpathDiff/ClasspathChangesComputerTest.kt [hunk 4]
```diff
@@ -27,6 +27,9 @@ abstract class ClasspathChangesComputerTest : ClasspathSnapshotTestCommon() {
     @Test
     abstract fun testModifiedAddedRemovedElements()
 
+    @Test
```
Comment (gavra0) [inline compiler/incremental-compilation-impl/test/org/jetbrains/kotlin/incremental/classpathDiff/ClasspathChangesComputerTest.kt:30]: Can you please add a test for the mixed granularity? E.g:
- class X with previous fine-grained and current coarse-grained
- prev_X(fine) & current_X(fine) and also prev_Y (coarse) & current_Y(coarse)
this is to make sure diffing logic can differentiate the two and that it is also able to sum of changes produced by these 2 different methods.

Comment (hungvietnguyen) [reply compiler/incremental-compilation-impl/test/org/jetbrains/kotlin/incremental/classpathDiff/ClasspathChangesComputerTest.kt:34]: Done - I've added it in the next commit.

File: compiler/incremental-compilation-impl/test/org/jetbrains/kotlin/incremental/classpathDiff/ClasspathSnapshotTestCommon.kt [hunk 5]
```diff
@@ -126,11 +128,19 @@ abstract class ClasspathSnapshotTestCommon {
             }
 
             val classesDir = tmpDir.newFolder()
-            org.jetbrains.kotlin.test.MockLibraryUtil.compileKotlin(
+            // Note: Calling the following is simpler:
+            //     org.jetbrains.kotlin.test.MockLibraryUtil.compileKotlin(
+            //         srcDir.path, classesDir, extraClasspath = classpath.map { it.path }.toTypedArray())
+            // However, it currently fails with UnsupportedClassVersionError, so we have to launch a new kotlinc process instead
+            // (Linux only).
```
[the remark this answers is not in the document]
Comment (hungvietnguyen) [reply compiler/incremental-compilation-impl/test/org/jetbrains/kotlin/incremental/classpathDiff/ClasspathSnapshotTestCommon.kt:135]: This code is only temporary, ideally we should use `K2JVMCompiler` directly and compile Kotlin files on the fly instead of checking in the class files.

Right now it is not possible (as I described in the comments), so this is just a quick hack for me for compile the Kotlin files locally (it's not run on the server).

I can make this work on Windows if you think it's important?

[the remark this answers is not in the document]
Comment (gavra0) [reply compiler/incremental-compilation-impl/test/org/jetbrains/kotlin/incremental/classpathDiff/ClasspathSnapshotTestCommon.kt:135]: It is ok to check it in as-is, but pls add an explicit check that throws if the platform is unsupported.

**Review by gavra0 — APPROVED** [review]
LGTM

File: compiler/incremental-compilation-impl/test/org/jetbrains/kotlin/incremental/classpathDiff/ClasspathSnapshotTestCommon.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (hungvietnguyen) [reply compiler/incremental-compilation-impl/test/org/jetbrains/kotlin/incremental/classpathDiff/ClasspathSnapshotTestCommon.kt]: Done - I've added support for Windows.

**Comment by hungvietnguyen** [comment]
This PR has been merged.
